pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,220 @@
+}
+
+type gitClient interface {
+	Clone(ctx context.Context, repoID, remote, branch, destination string) (git.Repo, error)
+}
+
+type applicationLister interface {
+	List() []*model.Application
+}
+
+type Reporter struct {
+	apiClient         apiClient
+	gitClient         gitClient
+	applicationLister applicationLister
+	config            *config.PipedSpec
+	gitRepos          map[string]git.Repo
+	gracePeriod       time.Duration
+	// Cache for the last fetched commit for each repository.
+	lastFetchedCommitCache cache.Cache
+	logger                 *zap.Logger
+}
+
+func NewReporter(
+	apiClient apiClient,
+	gitClient gitClient,
+	appLister applicationLister,
+	cfg *config.PipedSpec,
+	gracePeriod time.Duration,
+	logger *zap.Logger,
+) (*Reporter, error) {
+	cache, err := memorycache.NewLRUCache(defaultLastFetchedCommitCacheSize)
+	if err != nil {
+		return nil, err
+	}
+	return &Reporter{
+		apiClient:              apiClient,
+		gitClient:              gitClient,
+		applicationLister:      appLister,
+		config:                 cfg,
+		gracePeriod:            gracePeriod,
+		lastFetchedCommitCache: cache,
+		logger:                 logger.Named("app-config-reporter"),
+	}, nil
+}
+
+func (r *Reporter) Run(ctx context.Context) error {
+	r.logger.Info("start running app-config-reporter")
+
+	// Pre-clone to cache the registered git repositories.
+	r.gitRepos = make(map[string]git.Repo, len(r.config.Repositories))
+	for _, repoCfg := range r.config.Repositories {
+		repo, err := r.gitClient.Clone(ctx, repoCfg.RepoID, repoCfg.Remote, repoCfg.Branch, "")
+		if err != nil {
+			r.logger.Error("failed to clone repository",
+				zap.String("repo-id", repoCfg.RepoID),
+				zap.Error(err),
+			)
+			return err
+		}
+		r.gitRepos[repoCfg.RepoID] = repo
+	}
+
+	// FIXME: Think about sync interval of app config reporter
+	ticker := time.NewTicker(r.config.SyncInterval.Duration())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := r.checkApps(ctx); err != nil {
+				r.logger.Error("failed to check application configurations defined in Git", zap.Error(err))
+			}
+		case <-ctx.Done():
+			r.logger.Info("app-config-reporter has been stopped")
+			return nil
+		}
+	}
+}
+
+// checkApps checks and reports two types of applications.
+// One is applications registered in Control-plane already, and another is ones that aren't registered yet.
+func (r *Reporter) checkApps(ctx context.Context) (err error) {
+	if len(r.gitRepos) == 0 {
+		r.logger.Info("no repositories were configured for this piped")
+		return
+	}
+
+	var (
+		unusedApps      = make([]*pipedservice.ApplicationConfiguration, 0)
+		appsToBeUpdated = make([]*pipedservice.ApplicationConfiguration, 0)
+		appsMap         = r.listApplications()
+	)
+	for repoID, repo := range r.gitRepos {
```

Comment:
We can break this function to be simpler ones. For example

``` golang
func scan(cox context.Context) error 
  for _, repo := range r.gitRepos {
     // 1. Fetch to update the local data of each repo to ensure they are up-to-date.
     // 2. Continue if there was no newly added commits.
     // 3. shouldScanRepos = append(shouldScanRepos, repo)
  }

  for _, repo := range shouldScanRepos {
     if err := scanRepo(cox context.Context, repo, previousHash, currentHash); err == nil {
        // mark the new hash to the map
     }
  }
}

func scanRepo(ctx context.Context, repo Repo, previousHash, currentHash string) error {
   ...
}
```

Reply:
Obviously more readable than before

Reply:
But to be honest, since we need to send all unregisteredApps each time, we have to scan all repositories anyway.

Reply:
>Piped sends all the latest configs every time whatever the repo is changed or not. The reason is the cache in control-plane is likely to be removed, and configs in Git is also likely to be removed

As mentioned in the PR description, I'm looking to letting Piped send all the latest configs every time whatever the repo is changed or not. The reasons are:
- There are no new commits to the repository, but the Control-plane cache may have already been removed.
- If you send only the configuration to be updated, you can't address the case that the file itself is deleted from the repo.

Reply:
I get your points.

Checking by the diff still be valid in those cases. Scanning the full repository can be done for only the first time and then just the diff from the second time. Of course, we should send the full list of "un-registered" ones and the changed "registered" ones to the control-plane.

About the deleted files I think we can ignore them since app deletion only be allowed from web.

Reply:
Ah, you mean to separate updating the repository from reading the files. Cool, I have no objection.

>About the deleted files I think we can ignore them since app deletion only be allowed from web.

Sorry for confusing you. I was afraid that even though the files of unregistered applications were deleted, they were still in the cache and would be suggested when registering an application.

Reply:
>  I was afraid that even though the files of unregistered applications were deleted, they were still in the cache and would be suggested when registering an application.

After having the list of changed files, we still have to read their content to get app_name, kind... So we can know that case at parsing time to handle as normal.

Reply:
Sorry, my words were possibly not enough. I was talking about unregistered applications.
Could you tell me about your thoughts on always sending ALL unregistered apps each time?

Reply:
> Could you tell me about your thoughts on always sending ALL unregistered apps each time?

Yes, I think that is ok, we have to send the full list of unregistered apps because they are saved in the cache.
We are on the same page about that topic.